### PR TITLE
Generate id's with custom length and letters

### PR DESCRIPTION
**Pull request by Ughuuu**
### 📓 Description

I want to be able to generate id's with smaller length(eg 4-5) and a shorter alphabet

### ☑️ Checklist

- [x] Documentation is up to date
- [?] Versions are bumped as needed

### 🔗 Related issues
https://github.com/foxssake/noray/issues/44

### CODE REVIEW

**Comment by Ughuuu** [comment]
Need to test it locally and build it, will wait for the docker change to be merged so it's easier to build.

File: .env.example [hunk 1]
```diff
@@ -1,3 +1,11 @@
+# Random generation ===========================================================
+# If you change this, check the collision probability https://zelark.github.io/nano-id-cc/
+NORAY_RANDOM_ID_LENGTH=21 # For 10 id/hour, 15 trillion years
```
Comment (elementbound) [inline .env.example:3]: Please use the following config keys to make both IDs configurable:
- `NORAY_OID_LENGTH`
- `NORAY_OID_CHARSET`
- `NORAY_PID_LENGTH`
- `NORAY_PID_CHARSET`

File: .env.example [hunk 2]
```diff
@@ -1,3 +1,11 @@
...
+# NORAY_RANDOM_ID_LENGTH=4 # For 10 id/hour, 2 days
+# NORAY_RANDOM_ID_LENGTH=5 # For 10 id/hour, 19 days
+# NORAY_RANDOM_ID_LENGTH=6 # For 10 id/hour, 155 days
+NORAY_RANDOM_ID_LETTERS=useandom-26T198340PX75pxJACKVERYMINDBUSHWOLF_GQZbfghjklqvwyzrict
```
Comment (elementbound) [inline .env.example:7]: Just curious, what's the reason behind the letter ordering?

File: src/hosts/host.entity.mjs [hunk 3]
```diff
@@ -2,7 +2,8 @@
 import * as net from 'node:net'
 import * as dgram from 'node:dgram'
 /* eslint-enable */
-import { nanoid } from 'nanoid'
+import { nanoid, customAlphabet } from 'nanoid'
```
Comment (elementbound) [inline src/hosts/host.entity.mjs:5]: Please change to `import * as nanoid from 'nanoid'`, I'd prefer to have it apparent in the code that it's a nanoid :slightly_smiling_face:

**Review by elementbound — COMMENTED** [review]
Thanks for the PR! Left some suggestions.

File: .env.example (hunk 2 above)
Comment (Ughuuu) [reply .env.example:7]: This is just the default the packages has, and it's like this for some compression optimisation.
Comment from: https://github.com/ai/nanoid/blob/main/url-alphabet/index.js#L1C1-L3C34
// This alphabet uses `A-Za-z0-9_-` symbols.
// The order of characters is optimized for better gzip and brotli compression.
// Same as in non-secure/index.js

Comment (elementbound) [reply .env.example]: Fair enough!

File: .env.example [hunk 4]
```diff
@@ -1,3 +1,13 @@
+# Random generation ===========================================================
+# If you change this, check the collision probability https://zelark.github.io/nano-id-cc/
+NORAY_OID_LENGTH=21 # For 10 id/hour, 15 trillion years
+# NORAY_RANDOM_ID_LENGTH=4 # For 10 id/hour, 2 days
```
Comment (elementbound) [inline .env.example:4]: Minor: please update the comments as well
```suggestion
# NORAY_OID_LENGTH=4 # For 10 id/hour, 2 days
# NORAY_OID_LENGTH=5 # For 10 id/hour, 19 days
# NORAY_OID_LENGTH=6 # For 10 id/hour, 155 days
```

File: src/config.mjs [hunk 5]
```diff
@@ -10,6 +11,16 @@ const env = process.env
 * Noray configuration type.
 */
 export class NorayConfig {
+  oid = {
+    length: env.NORAY_OID_LENGTH ?? 21,
```
Comment (elementbound) [inline src/config.mjs:15]: Please use the integer parser:
```suggestion
    length: integer(env.NORAY_OID_LENGTH) ?? 21,
```

File: src/config.mjs [hunk 6]
```diff
@@ -10,6 +11,16 @@ const env = process.env
...
+    charset: env.NORAY_OID_CHARSET ?? urlAlphabet
+  }
+
+  pid = {
+    length: env.NORAY_PID_LENGTH ?? 128,
```
Comment (elementbound) [inline src/config.mjs:20]: Please use the integer parser:
```suggestion
    length: integer(env.NORAY_PID_LENGTH) ?? 21,
```

File: src/hosts/host.entity.mjs [hunk 7]
```diff
@@ -46,8 +47,9 @@ export class HostEntity {
   */
   constructor (options) {
     options && Object.assign(this, options)
-
-    this.oid ??= nanoid()
-    this.pid ??= nanoid(128)
+    const customAlphabetOid = nanoid.customAlphabet(config.oid.charset, +config.oid.length)
```
Comment (elementbound) [inline src/hosts/host.entity.mjs:50]: The config is expected to be immutable during the application run, so these two can be moved to the top as two methods, for example: 

```js
const generateOID = nanoid.customAlphabet(config.oid.charset, +config.oid.length)
const generatePID = nanoid.customAlphabet(config.pid.charset, +config.pid.length)
```

**Comment by elementbound** [comment]
FYI this will be a "feat" commit as it adds a new feature. Please bump the minor component in package.json.

File: .env.example (hunk 4 above)
Comment (elementbound) [reply .env.example:4]: Sorry for unresolving but the comments still say `NORAY_RANDOM_ID_LENGTH` instead of `NORAY_OID_LENGTH`

File: package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@foxssake/noray",
-  "version": "1.3.2",
+  "version": "1.4.2",
```
Comment (elementbound) [inline package.json:3]: Minor bumps reset the patch component, so it goes from 1.3.2 to 1.4.0

File: src/hosts/host.entity.mjs [hunk 9]
```diff
@@ -46,8 +50,7 @@ export class HostEntity {
   */
   constructor (options) {
     options && Object.assign(this, options)
-
-    this.oid ??= nanoid()
-    this.pid ??= nanoid(128)
+    this.oid ??= generateOID(+config.oid.length)
```
Comment (elementbound) [inline src/hosts/host.entity.mjs:53]: The functions returned by `nanoid.customAlphabet` already include the length, so no need to pass them here.

Comment (Ughuuu) [reply src/hosts/host.entity.mjs:53]: If I don't it doesnt work for some reason and hangs.

Comment (Ughuuu) [reply src/hosts/host.entity.mjs:53]: Let me try again

Comment (Ughuuu) [reply src/hosts/host.entity.mjs]: Nvm, was bc I wasn't converting env var to int.

**Review by elementbound — APPROVED** [review]
lgtm, thanks for bearing with the comments!
